garp: Announce Pods with Gratuitous ARP

markpash opened the pull request:
garp: Announce Pods with Gratuitous ARP

This introduces a new feature that advertises Pod IPs on the L2 domain
using Gratuitous ARP packets.

When enabled, k8s Pod upsert and delete events are processed and GARP
packets are sent on the chosen interface, when required.

The EventProcessor cell introduced here is what processes the k8s Pod
events and maintains an internal state to make sure to only send GARP
packets when the Pod is created, or the IP is changed for some reason.
Pod deletion events simply erase the entry from the state.

There are new agent flags and helm values introduced to enable the
feature and to chose which interface to send GARP packets on.

Signed-off-by: Mark Pashmfouroush <mark@isovalent.com>
```release-note
Added Gratuitous ARP Pod Announcements
```

joamaki commented on pkg/datapath/garp/events.go:
```diff
@@ -0,0 +1,180 @@
+	nodeTypes "github.com/cilium/cilium/pkg/node/types"
+)
+
+var nodePodsCell = cell.Provide(func(lc hive.Lifecycle, c k8sClient.Clientset) (resource.Resource[*corev1.Pod], error) {
```
Use `ProvidePrivate` for internal resources. It doesn't make sense for other modules to use this. Longer term we'll provide a shared `Resource[*Pod]` but that's right now stuck with solving the kvstore complication (on kvstore connect we switch from watching all pods to just local pods).

joamaki commented on pkg/datapath/garp/events.go:
```diff
@@ -0,0 +1,180 @@
+// determines if a Gratuitous ARP packet needs to be sent.
+var EventProcessorCell = cell.Module(
+	"garp-event-processor",
+	"GARP event processor",
```
Use a descriptive title that someone unfamiliar with this can understand, e.g. "Sends gratuitous ARP packet for each local pod". The title is always associated with the module id, e.g. it'll look like "garp-event-processor: Sends gratuitous ARP packets for each local pod" (once we implement the module dependency diagram generation anyway)

joamaki commented on pkg/datapath/garp/events.go:
```diff
@@ -0,0 +1,180 @@
+}
+
+func (ep *eventProcessor) Start(hive.HookContext) error {
+	ep.wp = workerpool.New(1)
```
`pkg/hive/job` provides a nicer API for this now

markpash replied:
I'll use this in a followup PR, as we want to backport this.

joamaki commented on daemon/cmd/cells.go:
```diff
@@ -59,6 +60,9 @@ var (
 		cni.Cell,
 
+		// Gratuitous ARP event processor emits GARP packets on k8s pod creation events.
+		garp.EventProcessorCell,
```
Better place for this is at `pkg/datapath/cells.go`

joamaki commented on daemon/cmd/daemon_main.go:
```diff
@@ -1623,6 +1624,7 @@ type daemonParams struct {
 	HealthAPISpec        *healthApi.Spec
 	ServiceCache         *k8s.ServiceCache
 	ClusterMesh          *clustermesh.ClusterMesh
+	GARPEventProcessor   garp.EventProcessor `optional:"true"`
```
I don't see anything using this in daemon so this can be removed. Was this added here just to enforce the construction of `GARPEventProcessor`? See comment about using `Invoke` in `events.go` on how this should be done.

joamaki commented on pkg/datapath/garp/events.go:
```diff
@@ -0,0 +1,180 @@
+
+// EventProcessorCell processes k8s pod events for the current node and
+// determines if a Gratuitous ARP packet needs to be sent.
+var EventProcessorCell = cell.Module(
```
Consider putting this into `cell.go` and naming it `var Cell`. Move `config.go` contents into it as well and merge `Cell` from `garp.go` into it. Idea being that `cell.go` defines the "public" API of the module, e.g. the cell definition, exported interface types and the configuration. That way it's easy at a glance to see what's being provided here.

joamaki commented on pkg/datapath/garp/events.go:
```diff
@@ -0,0 +1,180 @@
+	Cell,
+	nodePodsCell,
+
+	cell.Provide(newEventProcessor),
```
Since the `EventProcessor` doesn't really provide an API for other modules we can use `cell.Invoke` instead. This will also make sure that it's strictly constructed.

Something along the lines of:

```
// cell.go:
var Cell = cell.Module(
  "garp",
  "Sends Gratuitous ARPs for local pods",

   cell.Config(Config{}),
   cell.Provide(newGARPSender),
   cell.Invoke(registerEventProcessor),
),

// events.go:
func registerEventProcessor(p eventProcessorParams) {
  ep := &eventProcessor{...}

  group := p.JobRegistry.NewGroup(...)
  group.Add(job.OneShot("garp-event-processor", ep.run))

  p.Lifecycle.Append(group)
}
```

julianwiedmann commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -1083,6 +1083,13 @@ data:
+  l2-announce-pods: "true"
+  garp-interface: {{ .Values.garp.interface | quote }}
+{{- end }}
+
```
Could you point a PR or similar here for `Dylan's L2 helm values`?

markpash replied:
His PR is merged and so my config values now closely align with his. :)

julianwiedmann commented on pkg/datapath/garp/events_test.go:
```diff
@@ -0,0 +1,179 @@
+		cfg.L2AnnouncePods = true
+	})
+
+	h.PrintObjects() // TODO(markpash): remove
```
Should this happen now? :)

julianwiedmann reviewed (approved):
Just a small comments, I trust Jussi's review on the hive/cell aspects.

Do you also have doc updates for this feature lined up? :pray:

markpash committed:
garp: Announce Pods with Gratuitous ARP

This introduces a new feature that advertises Pod IPs on the L2 domain
using Gratuitous ARP packets.

When enabled, k8s Pod upsert and delete events are processed and GARP
packets are sent on the chosen interface, when required.

The GARP cell introduced here is what processes the k8s Pod
events and maintains an internal state to make sure to only send GARP
packets when the Pod is created, or the IP is changed for some reason.
Pod deletion events simply erase the entry from the state.

There are new agent flags and helm values introduced to enable the
feature and to chose which interface to send GARP packets on.

Signed-off-by: Mark Pashmfouroush <mark@isovalent.com>

sayboras reviewed (approved):
helm changes lgtm :heavy_check_mark:

joamaki commented on pkg/datapath/garp/cells.go:
```diff
@@ -0,0 +1,61 @@
+	flags.Bool(EnableL2PodAnnouncements, def.EnableL2PodAnnouncements, "Enable announcing Pod IPs with Gratuitous ARP")
+}
+
+var localPodsCell = cell.ProvidePrivate(func(lc hive.Lifecycle, c k8sClient.Clientset) (resource.Resource[*corev1.Pod], error) {
```
I'm really sorry about this, but we now have `LocalPodResource` available from `daemon/k8s/resources.go`, so this can now be dropped in favor of it. The only changes you need are removing this and changing `resource.Resource[*corev1.Pod]` into `k8s.LocalPodResource` (package is `daemon/k8s`). Note that it's `slim_corev1.Pod` now. Hope it contains everything you need (if it doesn't, let's extend the slim pod). 

I'm still in the process of switching `pkg/k8s/watchers/pod.go` to use `LocalPodResource`, but should be done with that soon after which we should (in the normal use when we don't watch all pods) have only a single pod watcher in the whole agent.

markpash committed:
garp: Switch processor to use cilium EndpointManager

Signed-off-by: Mark Pashmfouroush <mark@isovalent.com>

joamaki reviewed (approved):
LGTM!

joestringer reviewed (approved):
Approving on behalf of @cilium/docs-structure . However, I would expect some additional user-facing documentation and perhaps e2e tests(?) to accompany the feature. Please do follow up on those.
